gke-e2e-cluster-create/0.1/gke-e2e-cluster-create.yaml

```diff
@@ -0,0 +1,107 @@
+  - name: kubeconfig
+    description: |
+      A workspace into which a kubeconfig file called `kubeconfig` will be written that will contain the information
+      required to access the cluster. The `kubeconfig` will expect to use gcloud to authenticate, so in order for it to
+      be used it must be run in a container which contains both kubectl and gcloud.
+  results:
+  - name: cluster-name
+    description: The name of the cluster that was created.
+  steps:
+  - name: gcloud
+    image: google/cloud-sdk:slim@sha256:27b2c22bf259d9bc1a291e99c63791ba0c27a04d2db0a43241ba0f1f20f4067f
+    script: |
+      UNIQUE_STR=$(head /dev/urandom | tr -dc a-z0-9 | head -c 10 ; echo '')
+      UNIQUE_NAME=$(params.identifier)-$UNIQUE_STR
+
+      # Configure gcloud to use the provided project and service account
+      gcloud auth activate-service-account --key-file=$(workspaces.gcp-service-account.path)/$(params.private-key-path)
+      gcloud config set project $(params.project-name)
+
+      # Create a network and a new cluster
+      gcloud compute networks create $UNIQUE_NAME --project $(params.project-name) --subnet-mode=auto
+      gcloud beta container clusters create \
+        --quiet \
+        --enable-autoscaling \
+        --scopes=cloud-platform \
+        --enable-basic-auth \
+        --no-issue-client-certificate \
+        --project=$(params.project-name) \
+        --cluster-version=$(params.cluster-version) \
+        --min-nodes=$(params.min-nodes) \
+        --max-nodes=$(params.max-nodes) \
+        --region=$(params.region) \
+        --machine-type=$(params.machine-type) \
+        --image-type=$(params.image-type) \
+        --num-nodes=1 \
+        --network=$UNIQUE_NAME\
+        $UNIQUE_NAME
+
+      # Write the kubeconfig for connecting to the new cluster to the provided workspace
+      KUBECONFIG=$(workspaces.kubeconfig.path)/kubeconfig gcloud beta container clusters get-credentials \
+        --project=$(params.project-name) \
+        --region=$(params.region) \
+        $UNIQUE_NAME
+
+      # Get the tag used for the instances created and use that to apply firewall rules to them
+      INSTANCE_TAG=$(gcloud compute instances list \
+        --project=$(params.project-name) \
+        --filter=metadata.cluster-name=$UNIQUE_NAME \
+        --limit=1 \
+        --format=get\(tags.items\) | tr -d '\n')
+
+      # This firewall rule allows the cluster to expose SSH, TCP services running on 80 or 8080,
+      # and services exposed via the NodePort default range (https://kubernetes.io/docs/concepts/services-networking/service/#nodeport)
+      gcloud compute firewall-rules create e2e-ports-$UNIQUE_STR \
+        --project=$(params.project-name) \
+        --network=$UNIQUE_NAME \
+        --allow=tcp:22,tcp:80,tcp:8080,tcp:30000-32767,udp:30000-32767 \
+        --target-tags=$INSTANCE_TAG
+
+      printf $UNIQUE_NAME > /tekton/results/cluster-name
```

Comment:
Oooh one other one - the cluster name by itself isn't actually that useful - you generally need the region/zone to access it as well. Would that make sense as a result too? Or I guess we maybe don't need it since it's a param.

Reply:
whatever you think is best! it feels a bit weird to replicated something that is passed in as a param but if you think it's worth it, ill do it!

(maybe one day we'll have some kind of "cluster info" result type that will have all of this info?)